File: Exámenes/2016 Febrero/Haskell 2016/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g 
    | length lista == length (vertices g) = True
    | otherwise = False
    where
        lista = [v | v <- vertices g, mod (degree g v) 2 == 0]


-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = mkGraphEdges lista (edges grafo1)
    where 
        grafo1 = deleteEdge g (v, u)
        lista = [v | v <- vertices g, degree grafo1 v /= 0]

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = aux g v0 [v0]
    where
        aux g v xs 
            |v == sucesor = (elim, sucesor:xs)
            |otherwise = aux g sucesor (sucesor:xs)
                where
                    sucesor = head (successors g v)
                    elim = remove g (v, sucesor)


-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys = ys
connectCycles (x:xs) (y:ys)
    | x == y = [x] ++ ys ++ xs
    | otherwise = [x] ++ connectCycles xs (y:ys)

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle
    |length list /= 0 = head list
    |otherwise = error "no tienen vertices en comun"
        where 
            list = [v | v <- vertices g, v `elem` cycle]

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g 
    |not (isEulerian g) = error "el grafo no es euleriano"
    |isEmpty g = []
    |length (vertices g) == 1 = [head (vertices g)]
    |otherwise = eulerianCycle' g (head (vertices g))
        where
            eulerianCycle' g v
                | isEmpty g = []
                | otherwise = connectCycles cycle (eulerianCycle' newGraph newVertex)
                    where
                        cycle = snd (extractCycle g v)
                        newGraph = fst (extractCycle g v)
                        newVertex = vertexInCommon newGraph cycle 

File: Exámenes/2016 Febrero/Java 2016/J2016/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name:
 * Student's group:
 *
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import dataStructures.list.*;
import dataStructures.set.HashSet;
import dataStructures.set.Set;

import java.util.Iterator;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        // TO DO
        boolean res = false;
        Set<V> vertices = new HashSet<>();
        for(V vert : g.vertices()){
            if(g.degree(vert) % 2 == 0){
                vertices.insert(vert);
            }
        }
        return vertices.size() == g.numVertices();
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        // TO DO
        g.deleteEdge(v, u);
        for(V vert : g.vertices()){
            if(g.degree(vert) == 0){
                g.deleteVertex(vert);
            }
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        // TO DO
        List<V> ciclo = new ArrayList<>();
        Set<V> sucesores = g.successors(v0);
        ciclo.append(v0);
        V sucesor = sucesores.iterator().next();
        V v = v0;
        while(!sucesor.equals(v0)){
            ciclo.append(sucesor);
            remove(g, v, sucesor);
            v = sucesor;
            sucesor = g.successors(v).iterator().next();
        }
        return ciclo;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
        if(xs.isEmpty()){
            xs = ys;
        }else{
            List<V> lista = new ArrayList<>();
            Iterator<V> it = xs.iterator();
            while(it.hasNext()){
                V v = it.next();
                lista.append(v);
                if(v==ys.get(0)){
                    for(V y : ys){
                        lista.append(y);
                    }
                }
            }
            xs = lista;
        }
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
    		// TO DO
            Iterator<V> it = cycle.iterator();
            V vert = null;

            while(it.hasNext()){
                V v= it.next();
                if(g.vertices().isElem(v)){
                    vert = v;
                }
            }
    		return vert;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
    		// TO DO
        if(!isEulerian(g)){
            System.out.println("Grafo no euleriano");
            return null;
        } else {
            V vert = g.vertices().iterator().next();
            List<V> ciclo = extractCycle(g, vert);
            while(!g.isEmpty()){
                V u = vertexInCommon(g, ciclo);
                List<V> aux = extractCycle(g, u);
                connectCycles(ciclo, aux);
            }
            return ciclo;
        }
    }
}
